csharp-basic-topics/RemoveIllegalCharsFromPathAndFilenames/RemoveIllegalCharsFromPathAndFilenames/Program.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System.Text.RegularExpressions;
+
+namespace CodeMazeGuides.CSBasic
+{
+    public class Program
+    {
+        static void Main(string[] args)
+        {
+            List<string> pathList = new List<string> 
+            {
+                "C://Thisisavalidpath",
+                "D://Thi*IsINVALID/<<",
+                "E://Users/ValidPath/docs",
+                "F://Users/Photos/Cats",
+                "G://Inv@7***lidPath",
+                "H://Test/This<IS>Ano<>therInvalidPath"
+            };
+            var convCheckResult = GetInvalidPaths(pathList);
+            var linqCheckResult = GetInvalidPathsLINQ(pathList);
+            var linqHeaderCheckResult = GetInvalidPathsLINQHeaderFormat(pathList);
+            var regexCheckResult = GetInvalidPathsRegEx(pathList);
+
+            List<string> validPaths = new List<string>();
+
+            foreach (var path in linqCheckResult)
+            {
+                validPaths.Add(ReplaceInvalidChars(path, ""));
+            }
+
+            Console.WriteLine("Valid paths are:");
+
+            foreach (var path in validPaths)
+            {
+                Console.WriteLine(path);
+            }
+        }
+
+        public static List<string> GetInvalidPaths(List<string> paths)
+        {
+            var invalidChars = Path.GetInvalidPathChars().Concat(Path.GetInvalidFileNameChars()).ToArray();
```

Comment:
Wouldn't be `Union()` instead of `Concat()` better? It can contain duplicates and would be therefore less performant when we duplicate the lookup.

Reply:
Better still would be a HashSet<char> that we initialize with the invalid path chars and then call set.UnionWith with the second set. Then the for loop could be changed to something like:
foreach(var path in paths)
{
  if (path.Any(c => invalid.Contains(c)) result.Add();
}

My gut feeling is that would be slightly more performant. We could go a step further in readability/clarity if we just changed the list creation and looping to:

`return paths.Where(p => p.Any(c => invalid.Contains(c)).ToList();`

I don't suspect the performance difference would be that significant (we can always test and be sure)